File: Data Structure & Algorithm/[ DS ] Priority Queue.java
public class Main {
	public static int n = 100;
    public static int[] src={9,8,1,2,5,4,7,6,3};

    public static void main(String[] args) {
    	Heap heap = new Heap();
        for (int i=0;i<src.length;i++)
			heap.push(src[i]);

        for (int i=0;i<src.length;i++) {
			System.out.print(heap.pop()+" ");
		}
    }

    public static class Heap {
    	int[] heap = new int[n];
    	int hCnt=0;

    	public Heap() {}

    	public void push (int x) {
    		heap[++hCnt]=x;
    		int idx = hCnt;
    		while (idx>1 && heap[idx/2]>heap[idx]) {
    			if (idx==1 || heap[idx/2]<heap[idx])
    				break;
    			int tmp = heap[idx/2];
    			heap[idx/2]=heap[idx];
    			heap[idx]=tmp;
    			idx/=2;
    		}
    	}

    	public int pop() {
    		int pop = heap[1];
    		heap[1] = heap[hCnt--];
    		int idx =1;
    		int next;
    		while (true) {
    			next = idx*2;
    			if (next<hCnt && heap[next]>heap[next+1])
    				next++;
    			if (next>hCnt || heap[next]>heap[idx])
    				break;
    			int tmp = heap[idx];
    			heap[idx]=heap[next];
    			heap[next]=tmp;
    			idx=next;
    		}
    		return pop;
    	}
    }
}

File: Data Structure & Algorithm/[ DS ] Priority Queue.py
class Heap:
    heap = [0]
    size = 0
    def __init__(self):
        heap = [0]
        size = 0

    def push(self, x):
        self.heap.append(x)
        self.size += 1
        idx = self.size
        while idx > 1 and self.heap[idx//2] > self.heap[idx]:
            if self.heap[idx//2] < self.heap[idx]:
                break
            temp = self.heap[idx//2]
            self.heap[idx//2] = self.heap[idx]
            self.heap[idx] = temp
            idx //= 2
    
    def pop(self):
        pop = self.heap[1]
        self.heap[1] = self.heap[size]
        self.size -= 1
        idx = 1
        next = -1
        while True:
            next = idx * 2
            if next < self.size and self.heap[next] > self.heap[next+1]:
                next += 1
            if next > self.size or self.heap[next] > self.heap[idx]:
                break
            temp = self.heap[idx]
            self.heap[idx] = self.heap[next]
            self.heap[next] = temp
            idx = next
        return pop

heap = Heap()
for a in [1,5,2,4,6,8,3]:
    heap.push(a)



